From d499f590a112948e7cbb1174dd4a6ffa46595d46 Mon Sep 17 00:00:00 2001
From: Jianhui Zhao <zhaojh329@gmail.com>
Date: Thu, 14 Nov 2024 09:57:28 +0800
Subject: [PATCH] feat(dns): add support lookup from `/etc/hosts`

Signed-off-by: Jianhui Zhao <zhaojh329@gmail.com>
---
 dns.lua | 39 +++++++++++++++++++++++++++++++++++++++
 1 file changed, 39 insertions(+)

diff --git a/dns.lua b/dns.lua
index b34eabe..8bbac97 100644
--- a/dns.lua
+++ b/dns.lua
@@ -438,6 +438,40 @@ local function query(s, id, req, nameserver)
     return parse_response(data, id)
 end
 
+local function name_from_hosts(qname, opts)
+    local typ = opts.type or M.TYPE_A
+
+    for line in io.lines('/etc/hosts') do
+        if line:sub(1, 1) ~= '#' and line ~= '' then
+            local fields = {}
+
+            for field in line:gmatch('%S+') do
+                fields[#fields + 1] = field
+            end
+
+            local address = fields[1]
+            local address_type
+
+            if socket.is_ipv4_address(address) then
+                address_type = M.TYPE_A
+            elseif socket.is_ipv6_address(address) then
+                address_type = M.TYPE_AAAA
+            end
+
+            if address_type == typ and #fields > 1 then
+                for i = 2, #fields do
+                    if fields[i] == qname then
+                        return {{
+                            type = typ,
+                            address = address
+                        }}
+                    end
+                end
+            end
+        end
+    end
+end
+
 --[[
     opts is an optional Table that supports the following fields:
     type: The current resource record type, possible values are 1 (TYPE_A), 5 (TYPE_CNAME),
@@ -470,6 +504,11 @@ function M.query(qname, opts)
 
     opts = opts or {}
 
+    local res = name_from_hosts(qname, opts)
+    if res then
+        return res
+    end
+
     local nameservers = {}
 
     for _, nameserver in ipairs(opts.nameservers or {}) do
-- 
2.34.1

