From 13559c66e7121142f20e6a929622b338aaba29dd Mon Sep 17 00:00:00 2001
From: Jianhui Zhao <zhaojh329@gmail.com>
Date: Mon, 6 May 2024 19:07:40 +0800
Subject: [PATCH] perf: improve random seed

Signed-off-by: Jianhui Zhao <zhaojh329@gmail.com>
---
 CMakeLists.txt |  2 +-
 eco.c          | 14 +++++++++++---
 2 files changed, 12 insertions(+), 4 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 020cb5e..5683831 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -26,7 +26,7 @@ if (NOT LUA53_INCLUDE_DIRS OR NOT LUA53_LIBRARIES)
     message(FATAL_ERROR "Liblua 5.3 is required.")
 endif()
 
-add_compile_options(-D_GNU_SOURCE -Os -Wall -Werror --std=gnu99 -fno-strict-aliasing)
+add_compile_options(-D_GNU_SOURCE -DLUA_USE_LINUX -Os -Wall -Werror --std=gnu99 -fno-strict-aliasing)
 
 # configure a header file to pass some of the CMake settings to the source code
 configure_file(${CMAKE_CURRENT_SOURCE_DIR}/config.h.in ${CMAKE_CURRENT_BINARY_DIR}/config.h)
diff --git a/eco.c b/eco.c
index a077b92..d383f33 100644
--- a/eco.c
+++ b/eco.c
@@ -3,6 +3,7 @@
  * Author: Jianhui Zhao <zhaojh329@gmail.com>
  */
 
+#include <sys/time.h>
 #include <stdlib.h>
 #include <lualib.h>
 #include <unistd.h>
@@ -681,6 +682,14 @@ static void show_usage(const char *progname)
         , progname);
 }
 
+static void set_random_seed()
+{
+    struct timeval t;
+
+    gettimeofday(&t, NULL);
+    srandom(t.tv_usec * t.tv_sec);
+}
+
 int main(int argc, char *const argv[])
 {
     struct ev_loop *loop = EV_DEFAULT;
@@ -691,13 +700,12 @@ int main(int argc, char *const argv[])
 
     signal(SIGPIPE, SIG_IGN);
 
+    set_random_seed();
+
     L = luaL_newstate();
 
     luaL_openlibs(L);
 
-    luaL_loadstring(L, "math.randomseed(os.time())");
-    lua_pcall(L, 0, 0, 0);
-
     luaL_loadstring(L,
         "table.keys = function(t)"
         "local keys = {}"
-- 
2.34.1

