From b058b523d0016c543472a5f03bb381c389b02ba6 Mon Sep 17 00:00:00 2001
From: Jianhui Zhao <zhaojh329@gmail.com>
Date: Thu, 25 Apr 2024 15:04:14 +0800
Subject: [PATCH] fix(http): Fix parse HTTP/1.0 in `recv_status_line`

Signed-off-by: Jianhui Zhao <zhaojh329@gmail.com>
---
 http/client.lua | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/http/client.lua b/http/client.lua
index 324100b..83eaca9 100644
--- a/http/client.lua
+++ b/http/client.lua
@@ -86,7 +86,7 @@ local function recv_status_line(sock, timeout)
         return nil, err
     end
 
-    local code, status = data:match('^HTTP/1.1 +(%d+) +([%w%p ]*)\r?$')
+    local code, status = data:match('^HTTP/1.[01] +(%d+) +([%w%p ]*)\r?$')
     if not code or not status then
         return nil, 'invalid http status line'
     end
-- 
2.34.1

