From a626b0e5adb167f0fcced69c7359dca58a8b2544 Mon Sep 17 00:00:00 2001
From: Jianhui Zhao <zhaojh329@gmail.com>
Date: Tue, 23 Jul 2024 09:30:46 +0800
Subject: [PATCH 1/3] feat(socket): bind `htonl`, `htons`, `ntohl`, `ntohs`

Signed-off-by: Jianhui Zhao <zhaojh329@gmail.com>
---
 socket.c | 32 ++++++++++++++++++++++++++++++++
 1 file changed, 32 insertions(+)

diff --git a/socket.c b/socket.c
index d4bdbfa..5877281 100644
--- a/socket.c
+++ b/socket.c
@@ -976,6 +976,34 @@ static int lua_if_indextoname(lua_State *L)
     return 1;
 }
 
+static int lua_htonl(lua_State *L)
+{
+    uint32_t n = luaL_checkinteger(L, 1);
+    lua_pushinteger(L, htonl(n));
+    return 1;
+}
+
+static int lua_htons(lua_State *L)
+{
+    uint16_t n = luaL_checkinteger(L, 1);
+    lua_pushinteger(L, htons(n));
+    return 1;
+}
+
+static int lua_ntohl(lua_State *L)
+{
+    uint32_t n = luaL_checkinteger(L, 1);
+    lua_pushinteger(L, ntohl(n));
+    return 1;
+}
+
+static int lua_ntohs(lua_State *L)
+{
+    uint16_t n = luaL_checkinteger(L, 1);
+    lua_pushinteger(L, ntohs(n));
+    return 1;
+}
+
 static const luaL_Reg funcs[] = {
     {"socket", lua_socket},
     {"is_ipv4_address", lua_is_ipv4_address},
@@ -986,6 +1014,10 @@ static const luaL_Reg funcs[] = {
     {"inet_pton", lua_inet_pton},
     {"if_nametoindex", lua_if_nametoindex},
     {"if_indextoname", lua_if_indextoname},
+    {"htonl", lua_htonl},
+    {"htons", lua_htons},
+    {"ntohl", lua_ntohl},
+    {"ntohs", lua_ntohs},
     {NULL, NULL}
 };
 
-- 
2.34.1

