From dd5db47bcee22292bc25bcb1bfabf5cb33f90346 Mon Sep 17 00:00:00 2001
From: Jianhui Zhao <zhaojh329@gmail.com>
Date: Tue, 9 Apr 2024 16:37:07 +0800
Subject: [PATCH 7/9] fix(socket): `ev_io` and `ev_timer` should be stoped in
 `close`.

Signed-off-by: Jianhui Zhao <zhaojh329@gmail.com>
---
 socket.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/socket.c b/socket.c
index 9c2b57b..b9fef42 100644
--- a/socket.c
+++ b/socket.c
@@ -738,6 +738,7 @@ static int lua_closed(lua_State *L)
 static int lua_sock_close(lua_State *L)
 {
     struct eco_socket *sock = luaL_checkudata(L, 1, ECO_SOCKET_MT);
+    struct ev_loop *loop = sock->eco->loop;
 
     if (sock->fd < 0)
         return 0;
@@ -750,6 +751,9 @@ static int lua_sock_close(lua_State *L)
             unlink(un.sun_path);
     }
 
+    ev_timer_stop(loop, &sock->tmr);
+    ev_io_stop(loop, &sock->io);
+
     close(sock->fd);
 
     sock->fd = -1;
-- 
2.34.1

