From d45f275d4faeff107c82c7e83b68f8b564333d33 Mon Sep 17 00:00:00 2001
From: Jianhui Zhao <zhaojh329@gmail.com>
Date: Mon, 8 Apr 2024 08:55:19 +0800
Subject: [PATCH 2/4] fix(examples): fix typo

Signed-off-by: Jianhui Zhao <zhaojh329@gmail.com>
---
 examples/netlink/ip-addr.lua | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/examples/netlink/ip-addr.lua b/examples/netlink/ip-addr.lua
index b4f157b..da74e06 100755
--- a/examples/netlink/ip-addr.lua
+++ b/examples/netlink/ip-addr.lua
@@ -11,7 +11,7 @@ if not ok then
 end
 
 local res, err = addr.get(ifname)
-if not ok then
+if not res then
     print('get fail:', err)
     return
 end
-- 
2.34.1

